app/workers/set_tenant_id_worker.rb

```diff
@@ -1,33 +1,43 @@
 # frozen_string_literal: true
 
 class SetTenantIdWorker < ApplicationJob
+  queue_as :low
 
-  def perform(provider)
-    provider.update_column(:master, false)
+  RELATIONS = ["backend_apis", "log_entries", "alerts"].freeze
 
-    [:backend_apis, :log_entries, :alerts].each do |relation|
-      provider.public_send(relation).where(tenant_id: nil).find_each do |instance|
-        ModelTenantIdWorker.perform_later(instance, provider.tenant_id)
+  def perform(provider, relations)
+    relations.each do |relation|
+      assoc = provider.public_send(relation)
+      assoc.select(:id).where(tenant_id: nil).find_in_batches(batch_size: 100) do |instances|
+        ModelTenantIdWorker.perform_later(assoc.klass, instances.map(&:id), provider.tenant_id)
       end
     end
   end
 
   class BatchEnqueueWorker < ApplicationJob
     unique :until_executed
+    queue_as :low
 
-    def perform(*)
-      Account.providers.where(master: nil).find_each do |provider|
-        SetTenantIdWorker.perform_later(provider)
+    def self.validate_params(*relations)
+      raise "you must pass relations to fix" if relations.empty?
```

Comment:
Maybe we need a test for these raised exceptions?...

Reply:
added